src/main/kotlin/stringcalculator/Number.kt

```diff
@@ -0,0 +1,21 @@
+package stringcalculator
+
+@JvmInline
+value class Number(val value: Int) {
+    constructor(value: String) : this(
+        requireNotNull(value.toIntOrNull()) {
+            throw IllegalArgumentException("Invalid value: $value required number formattable value")
+        }
+    )
```

Comment:
`Inline Class`와 `Secondary Constructor`를 사용하셨군요👍👍

한솔님도 잘 알고 계시겠지만, `requireNotNull`은 인자가 null이 아닌 경우 `non-null` 타입을 반환하는데요,

아래의 경우는 간단히 `elvis` 연산만으로도 표현할 수 있을 것 같습니다..!
```suggestion
 @JvmInline
 value class Number(val value: Int) {
     constructor(value: String) : this(
         value.toIntOrNull() ?: throw IllegalArgumentException("Invalid value: $value required number formattable value")
     )

```

Reply:
`non-null`타입을 반환하는 경우 `elvis` 연산으로 표현하는게 기존 방식보다 어떤 부분에서 나은지 궁금합니다!

Reply:
두 방법 모두 null check에 초점을 둔다면 큰 차이는 없습니다!
다만 저의 개인적인 생각으로는 `elvis` 연산을 활용하는게 더 읽기 쉽고 유연하다고 생각합니다.(예를 들어 `value`에 기본값이 추가되는 경우를 생각해보면요!)

다시 보다보니 한가지 더 이야기를 나눠보고 싶은 부분을 찾았는데요..! 
위 코드에선 생성자에서 두 가지 기능이 수행된다고 볼 수 있습니다.
 - String -> Int? 변환
 - null check 후 예외 처리
 
한솔님도 알고 계시겠지만, 코틀린은 인스턴스가 생성되기 전 초기화를 지원하는 [init](https://kotlinlang.org/docs/classes.html#constructors)을 지원하기 때문에 자바처럼 생성자에서 입력값을 검증하거나 초기화하지 않아도 됩니다.
대신 객체를 생성할 때 필요한 입력값을 다루거나 예외처리를 해야 하는 경우라면 [팩토리 함수](https://kt.academy/article/ek-factory-functions)를 활용하는 방법은 어떻게 생각하시나요?

Reply:
자바의 정적 팩토리 메서드 패턴을 말씀하시는 것 같은데, 위 같은 해당 보조 생성자의 경우에는 문자열을 받고 숫자로 변환이 필요하고, 그 과정에서 문제가 될 만한 요지(blank or not number)들에 대한 예외처리를 해주었는데, 

제 개인적으로는 해당 로직의 비용이나 성능 부분에서 고려하거나 사이드 이펙트가 날 부분이 제가 직접만든 로직이 아닌 코틀린에서 제공하는 메서드(toIntOrNull)을 사용하기에 적다고 생각하여 별도의 팩토리 함수를 사용하지는 않았습니다. 

저는 개인적으로 다형성, 직접만든 커스텀 로직이 하나 이상 수행되야하는 경우에만 정적 팩토리 메서드 패턴을 사용하려 합니다 !